table.go

```diff
@@ -1391,3 +1072,85 @@ func (t *Table) close() {
 	t.active.pendingWritersWg.Wait()
 	t.closing = true
 }
+
+// configureLSMLevels configures the level configs for this table.
+func (t *Table) configureLSMLevels(levels []*LevelConfig) []*LevelConfig {
+	config := make([]*LevelConfig, 0, len(levels))
+
+	for _, level := range levels {
+		config = append(config, &LevelConfig{
+			Level:   level.Level,
+			MaxSize: level.MaxSize,
+			Compact: t.parquetCompaction, // NOTE: right now we only support parquet compaction but in the future this could compact parts into larger arrow records.
+		})
+	}
+
+	// Don't set the compaction functino for the last level. As the LSM doesn't have a level after the last one.
+	config[len(levels)-1].Compact = nil
+
+	return config
+}
+
+func (t *Table) parquetCompaction(compact []*parts.Part) (*parts.Part, int64, int64, error) {
+	b := &bytes.Buffer{}
+	size, err := t.compactParts(b, compact)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
+	buf, err := dynparquet.ReaderFromBytes(b.Bytes())
+	if err != nil {
+		return nil, 0, 0, err
+	}
+	return parts.NewPart(0, buf), size, int64(b.Len()), nil
+}
+
+func (t *Table) externalParquetCompaction(writer io.Writer) func(compact []*parts.Part) (*parts.Part, int64, int64, error) {
+	return func(compact []*parts.Part) (*parts.Part, int64, int64, error) {
+		size, err := t.compactParts(writer, compact)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+
+		return nil, size, 0, nil
+	}
+}
+
+// compactParts will compact the given parts into a Parquet file written to w.
+// It returns the size in bytes of the compacted parts.
+func (t *Table) compactParts(w io.Writer, compact []*parts.Part) (int64, error) {
+	bufs := []dynparquet.DynamicRowGroup{}
+	var size int64
+	for _, part := range compact {
```

Comment:
One thing we should definitely make sure to keep from the previous compaction is to consider non-overlapping parts a single record to merge, otherwise this is going to have very bad memory usage.

Reply:
Yes agreed, on my list to copy back over but thanks for the reminder